src/app/clusters/ecosystem-information-server/ecosystem-information-server.cpp

```diff
@@ -0,0 +1,368 @@
+{
+    Structs::HomeLocationStruct::Type homeLocationStruct;
+    // This would imply data is either not properly validated before being
+    // stored here or corruption has occurred.
+    VerifyOrDie(!aHomeLocation.mLocationName.empty());
+    homeLocationStruct.locationName = CharSpan(aHomeLocation.mLocationName.c_str(), aHomeLocation.mLocationName.size());
+
+    if (aHomeLocation.mFloorNumber.has_value())
+    {
+        homeLocationStruct.floorNumber.SetNonNull(aHomeLocation.mFloorNumber.value());
+    }
+    else
+    {
+        homeLocationStruct.floorNumber.SetNull();
+    }
+
+    if (aHomeLocation.mAreaType.has_value())
+    {
+        homeLocationStruct.areaType.SetNonNull(aHomeLocation.mAreaType.value());
+    }
+    else
+    {
+        homeLocationStruct.areaType.SetNull();
+    }
+    return homeLocationStruct;
+}
+
+} // namespace
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::SetDeviceName(std::string aDeviceName,
+                                                                               uint64_t aDeviceNameLastEditEpochUs)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mDeviceName                = std::move(aDeviceName);
+    mDeviceNameLastEditEpochUs = aDeviceNameLastEditEpochUs;
+    return *this;
+}
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::SetBrigedEndpoint(EndpointId aBridgedEndpoint)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mBridgedEndpoint = aBridgedEndpoint;
+    return *this;
+}
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::SetOriginalEndpoint(EndpointId aOriginalEndpoint)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mOriginalEndpoint = aOriginalEndpoint;
+    return *this;
+}
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::AddDeviceType(Structs::DeviceTypeStruct::Type aDeviceType)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mDeviceTypes.push_back(std::move(aDeviceType));
+    return *this;
+}
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::AddUniqueLocationId(std::string aUniqueLocationId,
+                                                                                     uint64_t aUniqueLocationIdsLastEditEpochUs)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mUniqueLocationIds.push_back(std::move(aUniqueLocationId));
+    mUniqueLocationIdsLastEditEpochUs = aUniqueLocationIdsLastEditEpochUs;
+    return *this;
+}
+
+std::unique_ptr<EcosystemDeviceStruct> EcosystemDeviceStruct::Builder::Build()
+{
+    bool deviceNameIsInvalid            = mDeviceName.size() > kDeviceNameMaxSize;
+    bool originalEndpointIsInvalid      = mOriginalEndpoint == kInvalidEndpointId;
+    bool deviceTypesListIsInvalid       = mDeviceTypes.empty();
+    bool uniqueLocationIdsListIsInvalid = mUniqueLocationIds.size() > kUniqueLocationIdsListMaxSize;
+    if (mIsAlreadyBuilt || deviceNameIsInvalid || originalEndpointIsInvalid || deviceTypesListIsInvalid ||
+        uniqueLocationIdsListIsInvalid)
+    {
+        return nullptr;
```

Comment:
here we lose the reason of why building failed. Since this is likely infrequently hit, would it make sense to log some error text whenever we return nullptr?

Also why do we do all boolean setting and only then an if? The first invalid should already fail like:

```
VerifyOrReturnValue(mDeviceName.size() <= kDeviceNameMaxSize, nullptr, ChipLogError(Zcl, "Device name too large"));
VerifyOrReturnValue(mOriginalEndpoint != kInvalidEndpointId;, nullptr, ChipLogError(Zcl, "Invalid original endpoint"));

....
```

Reply:
Thanks!